Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c

```diff
@@ -314,6 +316,220 @@ InternalEfiShellStartCtrlSMonitor(
+    &gEfiLoadedImageProtocolGuid,
+    (VOID**)&LoadedImage,
+    gImageHandle,
+    NULL,
+    EFI_OPEN_PROTOCOL_GET_PROTOCOL);
+
+  if (!EFI_ERROR(Status)) {
+    //
+    // If the image is not an app abort it.
+    //
+    if (LoadedImage->ImageCodeType != EfiLoaderCode){
+      ShellPrintHiiEx(
+        -1,
+        -1,
+        NULL,
+        STRING_TOKEN (STR_SHELL_IMAGE_NOT_APP),
+        ShellInfoObject.HiiHandle
+      );
+      goto UnloadImage;
+    }
+
+    ASSERT(LoadedImage->LoadOptionsSize == 0);
+    if (NewCmdLine != NULL) {
+      LoadedImage->LoadOptionsSize  = (UINT32)StrSize(NewCmdLine);
+      LoadedImage->LoadOptions      = (VOID*)NewCmdLine;
+    }
+
+    //
+    // Save our current environment settings for later restoration if necessary
+    //
+    if (Environment != NULL) {
+      Status = GetEnvironmentVariableList(&OrigEnvs);
+      if (!EFI_ERROR(Status)) {
+        Status = SetEnvironmentVariables(Environment);
+      }
+    }
+
+    //
+    // Initialize and install a shell parameters protocol on the image.
+    //
+    ShellParamsProtocol.StdIn   = ShellInfoObject.NewShellParametersProtocol->StdIn;
+    ShellParamsProtocol.StdOut  = ShellInfoObject.NewShellParametersProtocol->StdOut;
+    ShellParamsProtocol.StdErr  = ShellInfoObject.NewShellParametersProtocol->StdErr;
+    Status = UpdateArgcArgv(&ShellParamsProtocol, NewCmdLine, Efi_Application, NULL, NULL);
+    ASSERT_EFI_ERROR(Status);
+    //
+    // Replace Argv[0] with the full path of the binary we're executing:
+    // If the command line was "foo", the binary might be called "foo.efi".
+    // "The first entry in [Argv] is always the full file path of the
+    //  executable" - UEFI Shell Spec section 2.3
+    //
+    ImagePath = EfiShellGetFilePathFromDevicePath ((EFI_DEVICE_PATH_PROTOCOL*)DevicePath);
+    // The image we're executing isn't necessarily in a filesystem - it might
+    // be memory mapped. In this case EfiShellGetFilePathFromDevicePath will
+    // return NULL, and we'll leave Argv[0] as UpdateArgcArgv set it.
+    if (ImagePath != NULL) {
+      if (ShellParamsProtocol.Argv == NULL) {
+        // Command line was empty or null.
+        // (UpdateArgcArgv sets Argv to NULL when CommandLine is "" or NULL)
+        ShellParamsProtocol.Argv = AllocatePool (sizeof (CHAR16 *));
+        if (ShellParamsProtocol.Argv == NULL) {
+          Status = EFI_OUT_OF_RESOURCES;
+          goto UnloadImage;
+        }
+        ShellParamsProtocol.Argc = 1;
+      } else {
+        // Free the string UpdateArgcArgv put in Argv[0];
+        FreePool (ShellParamsProtocol.Argv[0]);
+      }
+      ShellParamsProtocol.Argv[0] = ImagePath;
+    }
+
+    Status = gBS->InstallProtocolInterface(&NewHandle, &gEfiShellParametersProtocolGuid, EFI_NATIVE_INTERFACE, &ShellParamsProtocol);
+    ASSERT_EFI_ERROR(Status);
+
+    ///@todo initialize and install ShellInterface protocol on the new image for compatibility if - PcdGetBool(PcdShellSupportOldProtocols)
```

Comment:
@krystian-hebel when we plan to implement that? If trampoline is fully functional without it I would recommend to put that into future improvements section.